README.md
# Backend-client

## Under consideration to be deprecated!

The current architecture of the backend client does not support easy addition of new features and there are quite many stability issues. Based on the initial analysis, major refactoring would be needed to make Backend client to support all the current use cases reliably. Hence, we are considering to break current backend client into focused functionalities that have value for wider community. This may imply that existing functionality is not supported anymore or it is supported in a new repository. More on this as work proceeds.









[![Codacy Badge](https://api.codacy.com/project/badge/Grade/bf0c23237cf04ea6ac48e98cf10b6a7b)](https://www.codacy.com/manual/wirepas/backend-client?utm_source=github.com&utm_medium=referral&utm_content=wirepas/backend-client&utm_campaign=Badge_Grade) [![Build Status](https://travis-ci.com/wirepas/backend-apis.svg?branch=master)](https://travis-ci.com/wirepas/backend-client)  [![Documentation Status](https://readthedocs.org/projects/backend-client/badge/?version=latest)](https://backend-client.readthedocs.io/en/latest/?badge=latest) [![PyPi](https://img.shields.io/pypi/v/wirepas-backend-client.svg)](https://pypi.org/project/wirepas-backend-client/)

<!-- MarkdownTOC levels="1,2,3" autolink="true"  style="ordered"  -->

1. [Introduction](#introduction)
1. [Installation](#installation)
    1. [Host dependencies](#host-dependencies)
    1. [Setting up a Python virtual environment](#setting-up-a-python-virtual-environment)
    1. [Installing from PyPi](#installing-from-pypi)
    1. [Installing from Github](#installing-from-github)
1. [Entrypoints](#entrypoints)
    1. [Gateway command line interface](#gateway-command-line-interface)
    1. [WPE Viewer](#wpe-viewer)
    1. [WNT Viewer](#wnt-viewer)
    1. [Provisioning server](#provisioning-server)
1. [Framework](#framework)
    1. [Structure](#structure)
    1. [Examples](#examples)
1. [Logging to fluentd](#logging-to-fluentd)
1. [Building and running over Docker](#building-and-running-over-docker)
    1. [Dockerhub](#dockerhub)
    1. [Running with docker](#running-with-docker)
    1. [Running with compose](#running-with-compose)
    1. [Building the image locally](#building-the-image-locally)
1. [Source documentation](#source-documentation)
1. [Contributing](#contributing)
1. [License](#license)

<!-- /MarkdownTOC -->

## Introduction

The Backend Client is a tool to serve as a decoder and a test framework
where you can extend the existing interfaces to develop your own test cases.
Interactions with
a [gateway][wirepas_gateway] must be compliant with the Wirepas
Backend to Gateway API.

In addition to the interaction with the Backend to Gateway Public API,
the Backend Client has also support for Wirepas Positioning Engine and Wirepas
Network Tool. Support for these tools is still work in progress
(see [milestones][backend_client_gh_milestones] and [projects][backend_client_gh_projects]).
However, you can already benefit from existing
entrypoints to help you capture data and debug your
integration (see [entrypoints](#entrypoints)).

Figure 1 illustrates the horizon where Backend Client is located.

![Architecture overview][here_img_overview]

**Figure 1** - Backend Client role in the Wirepas ecosystem.

## Installation

The Backend Client is written in python and it requires a valid MQTT
broker target to connect to, WNT or WPE credentials.

If you need help installing a MQTT broker or auxiliary software please
refer to our tutorials repository.

### Host dependencies

The main requirements of Backend Client are:

-   Python 3.7
-   Pip (we recommend the latest available)
-   Linux system

Certain Backend Client's dependencies must be
compiled locally upon installation,
thus require the installation of specific development headers.

Under Debian, the extra requirements are met with:

```bash
    sudo apt-get install default-libmysqlclient-dev gcc python3.7-dev
    
```

Please be advised that the name of such packages might change depending on your
Linux system. Please refer to the Dockerfiles
under the [container folder][here_container] for more detailed information.

Windows native environments are not supported. For help running the tool
in Windows, please use the Linux subsystem, Docker or a virtual machine.

### Setting up a Python virtual environment

As a recommendation, it is good practice to install and run the
application on a virtual python environment. Such approach avoids
possible issues with conflicting package versions.

We recommend you use [pipenv][pipenv]:

```shell
    pip3 install pipenv
```

If you choose to install and run the Backend Client inside a virtual
environment, please activate the environment before moving with the next
steps.

For pipenv, a virtual environment is created and activated by:

```shell
    pipenv --three
    pipenv shell
```

Once you activate your environment your can safely install the python
package under it.

For more information please refer to the tool's webpage.

### Installing from PyPi

The Backend Client is available from
[PyPi][backend_client_pypi] and you can install the latest stable version
with:

```shell
    pip install wirepas-backend-client
```

If you wish to install a particular version please see the release history
from PyPi.

### Installing from Github

First of all, make sure to clone the repository using the https address.

Checkout the
[git tag that corresponds to the release version][backend_client_gh_releases]
you want to install and install the package with:

```shell
    pip3 install .
```

If you leave your branch checked out to the top of master, it is likely that
you might run into a few issues. If you find them, please be so kind to
raise them to us through an [issue or bug report][backend_client_gh_issues].

If you want to develop or patch a bug under your local environment,
you can install the package in development mode through:

```shell
    pip3 install -e .
```

When installed in development mode, changes to the source files will be
immediately visible.

## Entrypoints

After installation, the Backend Client exposes several
[shell entrypoints][here_setup_entrypoints] that
can be run from any location in your system. This section describes their
usage in more detail.

### Gateway command line interface

The Backend Client exposes an interactive gateway command line which is
meant to interact with a gateway and sink pair connected to a given
MQTT broker.

To start an interactive command line to interact with a broker located
in your _localhost_ environment type:

```shell
    wm-gw-cli
```

Connecting to a remote server requires you to specify the values as an
argument or by reading from a yaml settings file. To view all the
arguments, please issue the command:

```shell
    wm-gw-cli --help
```

#### Connecting to the MQTT broker

For the gateway client to function properly, it needs to establish a
connection to the same MQTT broker where one or multiple gateways
are connected to.

To do so, let's assume the following MQTT credentials are saved in
./examples/settings.yml:

```yaml
 # examples/settings.yml
 # Example on how to set the MQTT credentials towards a local broker
 # with basic authentication
 mqtt_hostname: 127.0.0.1
 mqtt_password: password
 mqtt_username: username
 mqtt_port: 1883 # defaults to 8883 (secure port)
 mqtt_force_unsecure: True  # defaults to False (secure)
```

To start the gateway against the MQTT broker specified in
./examples/settings.yml, issue the following command:

```shell
    wm-gw-cli --settings ./examples/settings.yml
```

Once the client launches, you will be greeted with:

```shell
    Welcome to the Wirepas Gateway Client cli!
    Connecting to mosquittouser@127.0.0.1:1883 (unsecure: True)

    Type help or ? to list commands

    Type ! to escape shell commands
    Use Arrow Up/Down to navigate your command history
    Use CTRL-D or bye to exit

    09:50.17 | wm-gw-cli >
```

#### Commands

The gateway client has a set of commands which you can view by typing ?
in the client's shell:

```shell
    wm-gw-cli > ?

    Documented commands (type help <topic>):
    ========================================
    EOF                     playback           set_loop_iterations
    bye                     q                  set_loop_timeout
    clear_offline_gateways  record             set_reply_greeting
    eof                     scratchpad_status  set_sink
    gateway_configuration   scratchpad_update  settings
    gateways                scratchpad_upload  shell
    help                    selection          sinks
    list                    send_data          toggle_byte_print
    ls                      set_app_config     toggle_pretty_print
    networks                set_config         track_devices
    nodes                   set_gateway        track_data_packets
```

For each available command you can get more help if you type
**?command** for example, to view how to send a datagram,
type in the shell:

```shell
    wm-gw-cli > ?send_data
```

When typing commands, you can use the up and down arrows to browse your
shell history. There is also support for running host shell commands, by
prefacing the command with **!**.

#### Viewing connected devices

Upon a MQTT connection, the Backend Client will start populating an
internal structure based on any incoming message. The topic is parsed
and the information about the network devices are constructed.

The operation happens as messages are published from the network. It
will take time to build a full map of the network, which will be faster
depending on your network's packets per second.

In case you wish to view periodically which devices you have on your network,
type in the following command:

```shell
    wm-gw-cli > ?track_devices
```

To exit the tracking loop, give a new line feed to the input stream and
it will go back to the main command shell.

#### Device settings

When sending a downstream datagram, it is necessary to obey the
Backend to Gateway API (WM-RM-128).
For that reason, it is mandatory to know what is the target
gateway identifier and in most cases the sink id as well.

##### Selecting the target gateway and sink

The wm-gw-cli provides an easy way to select the target gateway and sink. To
make your choice, please type the following commands:

```shell
    wm-gw-cli > set_gateway
    0 : 2485378023427 : GatewayState.ONLINE
    1 : 2485378023426 : GatewayState.OFFLINE
    Please enter your gateway selection [0]: 0
    wm-gw-cli > set_sink
    0 : 3806491:2485378023427:sink0
    Please enter your sink selection [0]: 0
```

When setting the gateway, the wm-gw-cli shows you the following information
_object index : gateway id : state._ Similarly, for the sink, the wm-gw-cli
presents the object _index : network : gateway id: sink id_.

To make your selection, type the desired object index and press enter.

Once your selection is done, your commands will target that gateway id
and sink id. If you wish to view what your current selection is, use the
command:

```shell
    wm-gw-cli > selection
```

:warning: **WARNING** :warning:

The wm-gw-cli prompt will update itself with the current gateway and sink id.
Throughout this document, that information has been edited out of the
prompt.

##### Set app config

Setting the app config requires that your gateway and sink are selected.
In case you have not done the selection prior to this command, the wm-gw-cli
will ask for them.

When setting the app config, you are setting three values on the target
sink:

-   Diagnostic interval: the interval of network diagnostics (0, 30, 60,
    300, 600, 1800)

-   App config: A payload of a given size distributed to all associated
    devices on the network;


CHANGELOG.md
# Changelog

## [v1.3.7-rc.3](https://github.com/wirepas/backend-client/tree/v1.3.7-rc.3) (2020-08-11)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.7-rc.2...v1.3.7-rc.3)

**Merged pull requests:**

- Style check tool for next RC. [\#345](https://github.com/wirepas/backend-client/pull/345) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.7-rc.2](https://github.com/wirepas/backend-client/tree/v1.3.7-rc.2) (2020-08-10)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.7-rc.1...v1.3.7-rc.2)

**Merged pull requests:**

- Fix CTRL+C [\#344](https://github.com/wirepas/backend-client/pull/344) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Remove kpi-tester [\#343](https://github.com/wirepas/backend-client/pull/343) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.7-rc.1](https://github.com/wirepas/backend-client/tree/v1.3.7-rc.1) (2020-08-04)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.6...v1.3.7-rc.1)

**Merged pull requests:**

- Merge machine readable log display format feature. [\#341](https://github.com/wirepas/backend-client/pull/341) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Disable HTTP ACK [\#340](https://github.com/wirepas/backend-client/pull/340) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.6](https://github.com/wirepas/backend-client/tree/v1.3.6) (2020-06-29)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.6-rc.5...v1.3.6)

## [v1.3.6-rc.5](https://github.com/wirepas/backend-client/tree/v1.3.6-rc.5) (2020-06-26)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.6-rc.3...v1.3.6-rc.5)

**Merged pull requests:**

- Fix backend client low performance when handling incoming messages 1000 items/sec [\#336](https://github.com/wirepas/backend-client/pull/336) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Add support for the tags per node statistic [\#334](https://github.com/wirepas/backend-client/pull/334) ([PeiChunyan](https://github.com/PeiChunyan))

## [v1.3.6-rc.3](https://github.com/wirepas/backend-client/tree/v1.3.6-rc.3) (2020-06-12)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.6-rc.2...v1.3.6-rc.3)

**Merged pull requests:**

- Fix memory leak caused by adding handle http request. [\#331](https://github.com/wirepas/backend-client/pull/331) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.6-rc.2](https://github.com/wirepas/backend-client/tree/v1.3.6-rc.2) (2020-06-09)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.6-rc.1...v1.3.6-rc.2)

**Merged pull requests:**

- Add Gatewack response check to HTTP.PY API commands. [\#330](https://github.com/wirepas/backend-client/pull/330) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.6-rc.1](https://github.com/wirepas/backend-client/tree/v1.3.6-rc.1) (2020-06-04)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.5...v1.3.6-rc.1)

**Closed issues:**

- Make Track Data packets output more readable [\#315](https://github.com/wirepas/backend-client/issues/315)
- Installing Backend Client on macOS [\#176](https://github.com/wirepas/backend-client/issues/176)
- Add support for querying WNT apis from cmd line [\#74](https://github.com/wirepas/backend-client/issues/74)
- Add examples on how to use the WNT websocket API [\#73](https://github.com/wirepas/backend-client/issues/73)
- Move message decoding to wirepas\_messaging \(backend apis\) [\#56](https://github.com/wirepas/backend-client/issues/56)
- Automatic sourcing of messages [\#33](https://github.com/wirepas/backend-client/issues/33)
- WM Remote API support [\#31](https://github.com/wirepas/backend-client/issues/31)
- Daemon loop [\#28](https://github.com/wirepas/backend-client/issues/28)
- Mesh Device missing methods [\#22](https://github.com/wirepas/backend-client/issues/22)
- Mesh Management redesign [\#21](https://github.com/wirepas/backend-client/issues/21)

**Merged pull requests:**

- Fix set config [\#328](https://github.com/wirepas/backend-client/pull/328) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Made printing of incoming data packets more user friendly.  [\#327](https://github.com/wirepas/backend-client/pull/327) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Fixed set app config [\#326](https://github.com/wirepas/backend-client/pull/326) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Fix missing update of child tables to MySQL. [\#325](https://github.com/wirepas/backend-client/pull/325) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))
- Print gateway configuration to screen formatted manner [\#324](https://github.com/wirepas/backend-client/pull/324) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

## [v1.3.5](https://github.com/wirepas/backend-client/tree/v1.3.5) (2020-06-02)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v1.3.5-rc.9...v1.3.5)

## [v1.3.5-rc.9](https://github.com/wirepas/backend-client/tree/v1.3.5-rc.9) (2020-05-29)

[Full Changelog](https://github.com/wirepas/backend-client/compare/v.1.3.5-rc.1...v1.3.5-rc.9)

**Closed issues:**

- Add Ping! command. [\#314](https://github.com/wirepas/backend-client/issues/314)
- Disable currently broken commands from CLI [\#309](https://github.com/wirepas/backend-client/issues/309)
- Implement global scratchpad\_update command that also ask nodes to take firmware into use. [\#306](https://github.com/wirepas/backend-client/issues/306)
- Implement scratchpad\_check command [\#303](https://github.com/wirepas/backend-client/issues/303)

**Merged pull requests:**

- Correct SQL update load number so that show fractions too. [\#323](https://github.com/wirepas/backend-client/pull/323) ([ErkkiSalonenW](https://github.com/ErkkiSalonenW))

wirepas_backend_client/cli/shell.py
"""
    Shell
    =====

    Contains shells to connect to backend services.

    .. Copyright:
        Copyright 2020 Wirepas Ltd under Apache License, Version 2.0.
        See file LICENSE for full license details.
"""

import datetime
import logging
import os
import readline
import select
import subprocess
import sys
import time
import queue
from threading import Lock
import threading
from threading import Timer
from wirepas_backend_client.api.mqtt import Topics
from wirepas_backend_client.cli.gateway import GatewayCliCommands


class GatewayShell(GatewayCliCommands):
    """
    GatewayShell

    Implements an interactive shell to talk with gateway services.

    Attributes:
        intro (str) : what is printed on the terminal upon initialisation
        prompt (str)

    """

    # pylint: disable=locally-disabled, invalid-name, too-many-arguments,
    # pylint: disable=locally-disabled, too-many-public-methods, unused-argument, too-many-function-args

    def __init__(
        self,
        shared_state,
        tx_queue,
        rx_queue,
        settings,
        data_queue=None,
        event_queue=None,
        timeout=10,
        histfile_size=1000,
        exit_signal=None,
        logger=None,
    ):

        self._prompt_base = "cli"
        self._prompt_format = "{} | {} > "
        self._reply_greeting = "<< message from"

        self._bstr_as_hex = True
        self._pretty_prints = False
        self._minimal_prints = False
        self._silent_loop = False
        self._max_queue_size = 1000
        self._max_data_queue_size = 1  # Data queues contains list messges

        self._file = None
        self._histfile = os.path.expanduser("~/.wm-shell-history")
        self._histfile_size = histfile_size

        self._tracking_loop_timeout = 1
        self._tracking_loop_iterations = float("Inf")
        self._raise_errors = False
        self._skip_replies = True

        super().__init__()

        self.settings = settings
        self.intro = self.intro.format(**self.settings.to_dict())
        self.prompt = self._prompt_format.format(
            datetime.datetime.now().strftime("%H:%M.%S"), self._prompt_base
        )

        self.request_queue = tx_queue
        self.response_queue = rx_queue
        self.data_queue = data_queue
        self.event_queue = event_queue

        self.wait_api_lock = Lock()
        self.mqtt_topics = Topics()
        self.exit_signal = exit_signal
        self.timeout = timeout
        self.logger = logger or logging.getLogger(__name__)

        self.device_manager = shared_state["devices"]
        self._shared_state = shared_state
        self._flush_lock = threading.Lock()

        self.start_data_event_queue_peroidic_flush_timer()

    def start_data_event_queue_peroidic_flush_timer(self):
        self.data_event_flush_timer_interval_sec: float = 1.0
        self._perioidicTimer = Timer(
            self.data_event_flush_timer_interval_sec,
            self.__data_event_perioid_flush_timeout,
        ).start()

    def __data_event_perioid_flush_timeout(self):
        with self._flush_lock:
            self._trim_queues()
            self.start_data_event_queue_peroidic_flush_timer()

    @staticmethod
    def time_format():
        """ The prompt time format """
        return datetime.datetime.now().strftime("%H:%M.%S")

    @staticmethod
    def consume_queue(q, block=False, timeout=0):
        """ Exhaust a given queue until it is empty """
        while not q.empty():
            yield q.get(block=block, timeout=timeout)

    @staticmethod
    def strtobytes(payload):
        """ Handles the conversion of payload strings to bytes """
        try:
            payload = bytes.fromhex(payload)
        except ValueError:
            payload = bytes(payload, "utf-8")
        return payload

    @staticmethod
    def is_match(message, field, value):
        """ Checks if the class has an attribute that matches the provided field """
        match = True
        try:
            if value:
                match = bool(value == getattr(message, field))
        except AttributeError:
            pass
        return match

    @staticmethod
    def is_valid(args):
        """
        Returns True when all mandatory arguments have been set
        (not None)
        """
        for value in args.values():
            if value is None:
                return False
        return True

    @staticmethod
    def parse(line):
        """ Splits the command line in argument and options (kwargs) """
        args = line.split(" ")
        options = dict()
        delete_index = list()
        for arg in args:
            if "=" in arg:
                option = arg.split("=")
                options[option[0]] = option[1]
                delete_index.append(args.index(arg))

        list(map(args.pop, sorted(delete_index, reverse=True)))

        return args, options

    @staticmethod
    def get_option(option, index, cast, args, kwargs):
        """ Obtains the argument based on its key or through its index """

        if option in kwargs:
            return cast(kwargs[option])

        try:
            return cast(args[index])
        except IndexError:
            return None

    def retrieve_args(self, line, options=None):
        """
        Retrieve args parses the line for the provided command options.

        Arguments:
            line (str): the command line
            options (dict): a dictionary with the expected options

        Return:
            A dictionary with the parsed arguments

        """
        args, kwargs = self.parse(line)
        index = 0
        params = dict()

        if options:
            for name, details in options.items():

                try:
                    value = self.get_option(
                        name,
                        index=index,
                        cast=details["type"],
                        args=args,
                        kwargs=kwargs,
                    )
                except ValueError:
                    value = None

                if value is None and "default" in details:
                    value = details["default"]

                params[name] = value
                index += 1

        return params

    def consume_response_queue(self):
        """ Exhausts the response queue """
        for message in self.consume_queue(self.response_queue):
            self.on_response_queue_message(message)
            yield message

    def get_message_from_response_queue(self):
        message = None
        if self.response_queue.empty() is not True:
            message = self.response_queue.get()
        return message

    def consume_data_queue(self):
        """ Exhausts the data queue """
        while self.get_messages_from_data_queue() is not None:
            pass

    def get_messages_from_data_queue(self) -> list:
        msgs = None
        # Data message rate is huge compared to others. Queue contains message
        # lists that are then processed
        if self.data_queue.empty() is False:
            msgs = self.data_queue.get()  # returns message list
        return msgs

    def consume_event_queue(self):
        """ Exhausts the event queue """
        for message in self.consume_queue(self.event_queue):
            self.on_event_queue_message(message)
            yield message

    def get_message_from_event_queue(self):
        message = None
        if self.event_queue.empty() is not True:
            message = self.event_queue.get()
        return message

    def _trim_queues(self):
        """ Trim queues ensures that queue size does not run too long"""
        try:
            if self.data_queue.qsize() > self._max_data_queue_size:
                self.consume_data_queue()

            if self.event_queue.qsize() > self._max_queue_size:
                self.consume_event_queue()

            if self.response_queue.qsize() > self._max_queue_size:
                self.consume_response_queue()
        except FileNotFoundError:
            sys.exit("Exiting")

    def _update_prompt(self):
        """ Method called to update the prompt command line.
            Parent classes should overload the on_update_prompt
        """
        super().on_update_prompt()

    def _tracking_loop(
        self, cb, timeout=None, iterations=None, silent=False, **kwargs
    ):
        """ Simple tracking loop for period cli tasks """

        if timeout is None:
            timeout = self._tracking_loop_timeout

        if iterations is None:
            iterations = self._tracking_loop_iterations

        n_iter = 0
        while n_iter < iterations:
            n_iter = n_iter + 1

            if not silent or self._silent_loop is True:
                print(
                    "#{} : {}".format(
                        n_iter, datetime.datetime.now().isoformat("T")
                    )
                )

            cb(**kwargs)

            i, _, _ = select.select([sys.stdin], [], [], timeout)
            if i:
                sys.stdin.readline().strip()
                return True

    def _print(self, reply, reply_greeting=None, pretty=None):
        """ Prettified reply """
        super().on_print(reply, reply_greeting=None, pretty=None)

    def wait_for_answer(self, device, request_message, timeout=30, block=True):
        """ Wait response to request_message. If response received, return it.
            If timeout, return None

            !Note this function will discard messages it gets from queue.
            """

        self.wait_api_lock.acquire()

        wait_start_time = time.perf_counter()

        if device is None:
            raise ValueError
        if request_message is None:
            raise ValueError

        message = None
        if timeout:
            response_good: bool = False

            while response_good is False:
                try:
                    queue_poll_time_sec: float = 0.1
                    message = self.response_queue.get(
                        block=block, timeout=queue_poll_time_sec
                    )
                    if str(message.gw_id) == str(device):
                        if int(message.req_id) == int(
                            request_message["data"].req_id
                        ):
                            response_good = True
                        else:
                            pass

                    else:
                        pass

                    if self.request_queue.empty():
                        # wait a bit to avoid busy loop when putting
                        # same message back and reading it again.
                        default_sleep_time: float = 0.1
                        time.sleep(default_sleep_time)

                except queue.Empty:
                    # keep polling
                    pass

                if time.perf_counter() - wait_start_time > timeout:
                    print(
                        "Error got no reply for {} in time. "
                        "Time waited {:.0f} secs.".format(
                            device, time.perf_counter() - wait_start_time
                        )
                    )
                    message = None
                    break

        self.wait_api_lock.release()
        return message

    def notify(self):
        """
        Notify sets the shared dictionary to propagate changes to the other
        processes
        """
        self._shared_state["devices"] = self.device_manager

    def disabled_do_toggle_print_minimal_information(self, line):
        """
        Switches the byte prints as hex strings or python byte strings
        """
        self._minimal_prints = not self._minimal_prints
        print("Minimal prints: {}".format(self._minimal_prints))

    def disabled_do_toggle_byte_print(self, line):
        """
        Switches the byte prints as hex strings or python byte strings
        """
        self._bstr_as_hex = not self._bstr_as_hex
        print("hex prints: {}".format(self._bstr_as_hex))

    def disabled_do_toggle_pretty_print(self, line):
        """
        Switches between json or pretty print
        """
        self._pretty_prints = not self._pretty_prints
        print("pretty prints: {}".format(self._pretty_prints))

    def disabled_do_toggle_silent_loop(self, line):
        """
        Enables/disables the tracking loop verbosity
        """
        self._silent_loop = not self._silent_loop
        print("track loop prints: {}".format(self._silent_loop))

    def disabled_do_toggle_raise_errors(self, line):
        """ Sets the raise error toggle """
        self._raise_errors = not self._raise_errors
        print("raise errors: {}".format(self._raise_errors))

    def disabled_do_set_loop_iterations(self, line):
        """
        Sets the amount of loop iterations
        """
        options = dict(
            iterations=dict(type=int, default=self._tracking_loop_iterations)
        )
        args = self.retrieve_args(line, options)

        self._tracking_loop_iterations = args["iterations"]
        print(
            "track loop iterations: {}".format(self._tracking_loop_iterations)
        )

    def disabled_do_set_loop_timeout(self, line):
        """
        Sets the loop evaluation time
        """
        options = dict(
            timeout=dict(type=int, default=self._tracking_loop_timeout)
        )
        args = self.retrieve_args(line, options)

        self._tracking_loop_timeout = args["timeout"]
        print("track loop timeout: {}".format(self._tracking_loop_timeout))

    def disabled_do_set_reply_greeting(self, line):
        """
        Sets the reply greeting
        """

        if line.lower() in ("", "none", " "):
            line = None

        self._reply_greeting = line
        print("reply greeting set to: {}".format(self._reply_greeting))

    def do_settings(self, line):
        """
        Prints outs the settings acquired upon starting
        """
        self._print(self.settings, reply_greeting="settings:")

    def do_bye(self, line):
        """
        Exits the cli

        Usage:
            bye
        """
        print("Thank you for using Wirepas Gateway Client")
        self.close()
        if not self.exit_signal.is_set():
            self.exit_signal.set()

        return True

    def do_q(self, line):
        """
        Exits the cli

        Usage:
            q
        """
        return self.do_bye(line)

    def disabled_do_eof(self, line):
        """ Captures CTRL-D """
        return self.do_bye(line)

    def disabled_do_EOF(self, line):
        """ Captures CTRL-D """
        return self.do_bye(line)

    def disabled_do_record(self, arg="shell-session.record"):
        """
        Saves typed commands in a file for later playback

        Usage:
            record [filename (default: shell-session.record)]
        """
        self.close()
        self._file = open(arg, "w")

    def disabled_do_playback(self, arg="shell-session.record"):
        """
        Plays commands from a file

        Usage:
            plaback [filename (default: shell-session.record)]
        """
        try:
            with open(arg) as f:
                lines = f.read().splitlines()
                if "bye" in lines[-1]:
                    del lines[-1]
                self.cmdqueue.extend(lines)
        except TypeError:
            print("wrong file name")

    def do_help(self, command, args=None):
        """ Prints the command help and the acquired arguments (if given)"""
        # pylint: disable=locally-disabled, arguments-differ
        if args is not None:
            print(f"insufficient/incorrect arguments: {args}")
        super().do_help(command)

    @staticmethod
    def do_shell(line):
        """ Escape shell commands with! """
        args = line.split()
        try:
            output = subprocess.check_output(args, shell=False)
            print(output.decode())
        except FileNotFoundError:
            print("Unknown shell command: {}".format(args))

    def emptyline(self):
        """ What happens when an empty line is provided """
        self.consume_response_queue()

    def precmd(self, line):
        """ Executes before a command is run in onecmd """
        with self._flush_lock:  # Either we flush queue or process command
            self.device_manager = self._shared_state["devices"]

            if (
                self._file
                and "playback" not in line
                and "bye" not in line
                and "close" not in line
            ):
                print(line, file=self._file)

            if self._shared_state["devices"]:
                self.consume_response_queue()
                self._update_prompt()

        return line

    def postcmd(self, stop, line):
        """ Method called after each command is executed """
        with self._flush_lock:  # Either we flush queue or process command
            if self._shared_state["devices"]:
                self._update_prompt()
        return stop

    def onecmd(self, line):
        """
            Executes each command, escaping it for errors and updates the
            prompt with the current time and selection identifiers.

        """
        # pylint: disable=locally-disabled, broad-except
        with self._flush_lock:  # Either we flush queue or process command
            if self.exit_signal.is_set():
                return self.do_bye(line)

            try:
                if self.device_manager:
                    return super().onecmd(line)
            except Exception as err:
                print("Something went wrong:{}".format(err))
                if self._raise_errors:
                    raise

        return False

    def preloop(self):
        """ runs before the cmd loop is started """
        if os.path.exists(self._histfile):
            readline.read_history_file(self._histfile)

    def postloop(self):
        """ runs when the cmd loop finishes """
        readline.set_history_length(self._histfile_size)
        readline.write_history_file(self._histfile)

    def close(self):
        """
        Terminates the playback recording
        """
        if self._file:
            self._file.close()
            self._file = None
